### PR TITLE
Role-based access management in course search

### PR DESCRIPTION
**Pull request by NEOGDAWG**
This PR adds role-based access management to Course Search so users only see courses they have access to.

- Added direct course access filtering using `course_users`.
- Added role-based access filtering using the existing `course_user_role` rows for Program Directors, Department Heads, and faculty-wide access.
- Added an administrator bypass so administrators can search all courses.
- Updated course and program filter options to only include accessible courses.
- Updated Program View so programs only appear when they contain an accessible course.
- Updated program links so inaccessible programs are displayed without a link.
- Applied the same access restrictions to indexed course material-content search.
- Added and updated tests for each access type and related edge cases.
- Added an optional performance seeder and documented the performance-testing results.
- Updated the search and role-access documentation.

Notes:

- Search controls course visibility, while existing permission levels still determine whether a user can view or edit a course.
- Role-based search access relies on the existing materialized `course_user_role` rows.
- Remaining access synchronization edge cases have been documented: See  #20 

Tests:

- Ran the complete Composer test suite: 166 tests passed with 518 assertions.
- Manually verified authenticated material-content searches for administrator and no-access accounts.

Please refer to `docs/search.md`, `docs/role-based-access-management.md`, and the performance-testing report for more details.

### CODE REVIEW

File: laravel/app/Http/Controllers/SearchController.php [hunk 1]
```diff
@@ -86,11 +90,21 @@ public function index(Request $request){
...
             ->get();
 
+        // Program page access can come from direct permissions or an elevated role.
+        $programPageAccessIds = $user->programs()
+            ->pluck('programs.program_id')
+            ->merge($user->programsWithElevatedRoleAccess()->pluck('programs.program_id'))
+            ->map(fn ($programId) => (int) $programId)
+            ->unique()
+            ->values()
+            ->all();
```
Comment (FabianLCH) [inline laravel/app/Http/Controllers/SearchController.php]: I think you can simplify down using the `allPrograms` method defined for `User` objects, which already merges direct-access and role-based-access programs:
```suggestion
        $programPageAccessIds = $user->allPrograms()->pluck('program_id');
```

File: laravel/database/seeders/SearchPerformanceSeeder.php
Comment (FabianLCH) [inline laravel/database/seeders/SearchPerformanceSeeder.php]: To make the code more readable, I'm thinking we can add a couple of comments here to explain the main strategies you used to deterministically generate courses, programs, topics, etc.

File: laravel/tests/Feature/SearchTest.php [hunk 2]
```diff
@@ -61,6 +64,352 @@ public function test_guest_user_is_redirected_from_search_page()
...
+        $this->assertSearchVisibility($response, [
+            'Directorium Visible Course',
+        ], [
+            'Directorium Hidden Course',
+        ]);
+        $response->assertSee('Courses: 1');
+    }
+
+    public function test_program_role_without_course_role_does_not_grant_search_access()
+    {
```
Comment (FabianLCH) [inline laravel/tests/Feature/SearchTest.php]: Though the test correctly verifies the current behaviour, I feel this should not be the case. A program director should have access to all courses attached to the program(s) they are the director of

File: laravel/tests/Feature/SearchTest.php
Comment (FabianLCH) [inline laravel/tests/Feature/SearchTest.php]: I think we can add a few tests here:

- all non-admins can see the only names of programs they only have direct access to (no courses) in the "program" view
- program directors of *multiple* programs can see results from all the corresponding programs
- program directors can see filters of programs they have both role-based AND direct access to

**Review by FabianLCH — CHANGES_REQUESTED** [review]
Mostly minor comments to address but the changes are looking good!

File: laravel/app/Http/Controllers/SearchController.php (hunk 1 above)
Comment (NEOGDAWG) [reply laravel/app/Http/Controllers/SearchController.php]: Updated this to use the existing allPrograms() method. I did not apply the suggestion directly because pluck() returns a Collection, while the existing Blade checks use in_array(), which requires a native array. I added "->all()" to preserve the current expected type.

File: laravel/database/seeders/SearchPerformanceSeeder.php
Comment (NEOGDAWG) [reply laravel/database/seeders/SearchPerformanceSeeder.php]: added!

File: laravel/tests/Feature/SearchTest.php
Comment (NEOGDAWG) [reply laravel/tests/Feature/SearchTest.php]: Added all three tests. The direct-program cases exposed that program names and filter options were only being included when the program contained an accessible course, even if the user had direct access to the program itself. I updated the shared program access query to include direct and role-based program access while keeping course results restricted by course access.

**Review by FabianLCH — APPROVED** [review]
We're good to go, thanks Kabir!
